contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,870 @@
+    bytes public constant REBASE_CHECK_NEAREST_EPOCH_DISTANCE_VALUE = hex"04";
+
+    string public constant REBASE_CHECK_DISTANT_EPOCH_DISTANCE_KEY = "REBASE_CHECK_DISTANT_EPOCH_DISTANCE";
+    bytes public constant REBASE_CHECK_DISTANT_EPOCH_DISTANCE_VALUE = hex"0a";
+
+    string public constant VALIDATOR_DELAYED_TIMEOUT_IN_SLOTS_KEY = "VALIDATOR_DELAYED_TIMEOUT_IN_SLOTS";
+    bytes public constant VALIDATOR_DELAYED_TIMEOUT_IN_SLOTS_VALUE = hex"1c20";
+
+    string public constant VALIDATOR_DELINQUENT_TIMEOUT_IN_SLOTS_KEY = "VALIDATOR_DELINQUENT_TIMEOUT_IN_SLOTS";
+    bytes public constant VALIDATOR_DELINQUENT_TIMEOUT_IN_SLOTS_VALUE = hex"5460";
+
+    string public constant PREDICTION_DURATION_IN_SLOTS_KEY = "PREDICTION_DURATION_IN_SLOTS";
+    bytes public constant PREDICTION_DURATION_IN_SLOTS_VALUE = hex"c4e0";
+
+    string public constant FINALIZATION_MAX_NEGATIVE_REBASE_EPOCH_SHIFTS_KEY = "FINALIZATION_MAX_NEGATIVE_REBASE_EPOCH_SHIFT";
+    bytes public constant FINALIZATION_MAX_NEGATIVE_REBASE_EPOCH_SHIFT_VALUE = hex"0546";
+
+    string public constant NODE_OPERATOR_NETWORK_PENETRATION_THRESHOLD_BP_KEY = "NODE_OPERATOR_NETWORK_PENETRATION_THRESHOLD_BP";
+    bytes public constant NODE_OPERATOR_NETWORK_PENETRATION_THRESHOLD_BP_VALUE = hex"64";
+
+    //
+    // STRUCTURED STORAGE
+    //
+    bool public isUpgradeStarted;
+    bool public isUpgradeFinished;
+
+    /// Need to be called before LidoOracle implementation is upgraded to LegacyOracle
+    function startUpgrade() external {
+        _startUpgrade();
+    }
+
+    function finishUpgrade() external {
+        _finishUpgrade();
+    }
+
+    /// Perform basic checks to revert the entire upgrade if something gone wrong
+    function assertUpgradeIsFinishedCorrectly() external view {
+        _assertUpgradeIsFinishedCorrectly();
+    }
+
+    function revertIfUpgradeNotEnacted() external view {
+        if (!isUpgradeFinished) {
+            revert UpgradeNotEnacted();
+        }
+    }
+
+    function _startUpgrade() internal {
+        if (msg.sender != _voting) revert OnlyVotingCanUpgrade();
+        if (isUpgradeStarted) revert CanOnlyStartOnce();
+        if (_lidoOracle.getVersion() != EXPECTED_FINAL_LEGACY_ORACLE_VERSION - 1) {
+            revert LidoOracleMustNotBeUpgradedToLegacyYet();
+        }
+        _assertInitialProxyImplementations();
+        isUpgradeStarted = true;
+
+        _locator.proxy__upgradeTo(_locatorImplementation);
+
+        _upgradeProxyImplementations();
+
+        // Need to have the implementations attached to the proxies at this point
+        _assertInitialACL();
+
+        _withdrawalVault.initialize();
+        _initializeWithdrawalQueue();
+        _initializeAccountingOracle();
+        _initializeValidatorsExitBus();
+        _initializeStakingRouter();
+
+        _migrateLidoOracleCommitteeMembers();
+        _migrateDSMGuardians();
+    }
```

Comment:
may reconsider contents arrangement for `_startUpdate` and `_finishUpdate` (can have only tiny-ish scope for `_startUpdate` regarding legacy oracle and anything else in `_finishUpdate`